# **Лабораторная работа №1. Первичный анализ набора данных (Pandas)**

# Этапы выполнения проекта.
Проект состоит из трех этапов:

1. Обзор данных
2. Предварительная обработка данных
3. Проверка гипотез

# Этап 1. Обзор данных

***Описание столбцов***:<br>
•	'color' — цветная гамма фильма(черно-белый/цветной)<br>
•	'Director_Name' — имя режиссёра<br>
•	'num_Critic_for_reviews' — количество отзывов критиков о фильме<br>
•	'duration' - длительность фильма<br>
•	'director_Facebook_likes' — Количество лайков на Facebook странице режиссера<br>
•	'actor_3_Facebook_likes' — количество лайков на странице Facebook актёра 3<br>
•	'actor_2_name' — имя второго актёра<br>
•	'Actor_1_Facebook_likes' — количество лайков на странице Facebook актёра 3 <br>
•	'gross' — валовой доход фильма в долларах<br>
•	'genres' — жанр<br>
•	'num_user_for_reviews' — количество пользователей, оставивших отзыв о фильме<br>
•	'language' — язык<br>
•	'country' — страна<br>
•	'content_rating' — возрастное ограничение фильма<br>
•	'budget' — бюджет фильма в долларах<br>
•	'title_year' — год выпуска фильма<br>
•	'actor_2_facebook_likes' — количество лайков на странице Facebook актёра 2<br>
•	'imdb_score' — рейтинг фильма на IMDB<br>
•	'aspect_ratio' — соотношение сторон, в которых фильм был снят<br>
•	'movie_facebook_likes' — количество лайков на Facebook странице фильма<br>



In [4]:
import pandas as pd
from bs4 import BeautifulSoup
import requests

In [2]:
df = pd.read_csv('../data/Dataset1.csv')

In [3]:
pd.set_option('display.max_rows', 101)
df.head(100)

color       Director_Name  \
0                                               Color       James Cameron   
1                                              Colour      Gore Verbinski   
2                                              Colour          Sam Mendes   
3                                               Color   Christopher Nolan   
4                                                 NaN         Doug Walker   
5                                              Colour      Andrew Stanton   
6                                               Color           Sam Raimi   
7                                               Color        Nathan Greno   
8                                              Colour         Joss Whedon   
9                                              Colour         David Yates   
10                                              Color         Zack Snyder   
11                                              Color        Bryan Singer   
12                                              Color        Marc Forster   
13                                              Color      Gore Verbinski   
14                                              Color      Gore Verbinski   
15                                              color         Zack Snyder   
16                                              Color      Andrew Adamson   
17                                              Color         Joss Whedon   
18                                              Color        Rob Marshall   
19                                              Color    Barry Sonnenfeld   
20                                              Color       Peter Jackson   
21                                              color           Marc Webb   
22                                              Color        Ridley Scott   
23                                              Color       Peter Jackson   
24                                              Color         Chris Weitz   
25                                              Color       Peter Jackson   
26                                              Color       James Cameron   
27                                              Color       Anthony Russo   
28                                              Color          Peter Berg   
29                                              Color     Colin Trevorrow   
30                                              Color          Sam Mendes   
31                                              Color           Sam Raimi   
32                                              Color         Shane Black   
33                                              Color          Tim Burton   
34                                              Color        Brett Ratner   
35                                              Color         Dan Scanlon   
36                                              Color         Michael Bay   
37                                              Color         Michael Bay   
38                                              Color           Sam Raimi   
39                                              Color           Marc Webb   
40                                              Color     Joseph Kosinski   
41                                              Color       John Lasseter   
42                                              Color     Martin Campbell   
43                                              Color         Lee Unkrich   
44                                              Color                 McG   
45                                              Color           James Wan   
46                                              Color        Marc Forster   
47                                              Color        Bryan Singer   
48                                              Color         J.J. Abrams   
49                                              Color        Bryan Singer   
50                                              Color        Baz Luhrmann   
51                                          

Получение общей информации о таблице:

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   color                       5025 non-null   object 
 1   Director_Name               4872 non-null   object 
 2   num_Critic_for_reviews      4927 non-null   float64
 3   duration                    4959 non-null   float64
 4   director_Facebook_likes     4872 non-null   float64
 5   actor_3_Facebook_likes      4953 non-null   float64
 6   actor_2_name                4963 non-null   object 
 7   Actor_1_Facebook_likes      4968 non-null   float64
 8   gross                       4104 non-null   float64
 9   genres                      4974 non-null   object 
 10  actor_1_name                4968 non-null   object 
 11  movie_Title                 4974 non-null   object 
 12  num_voted_users             4974 non-null   float64
 13   cast_total_facebook_likes  4974 

# **Этап 2. Предварительная обработка данных (Data preprocessing)**

Необходимо исправить названия столбцов так, чтобы все буквы были строчными и чтобы не было пробелов в названии.

In [5]:
df.columns

Index(['color', 'Director_Name', 'num_Critic_for_reviews', 'duration',
       'director_Facebook_likes', 'actor_3_Facebook_likes', 'actor_2_name',
       'Actor_1_Facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_Title', 'num_voted_users', ' cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes;'],
      dtype='object')

In [6]:
df = df.rename(
    columns = {
        'Director_Name': 'director_name',
        'num_Critic_for_reviews': 'num_critic_for_reviews',
        'director_Facebook_likes': 'director_facebook_likes',
        'actor_3_Facebook_likes': 'actor_3_facebook_likes',
        'Actor_1_Facebook_likes': 'actor_1_facebook_likes',
        'movie_Title': 'movie_title',
        ' cast_total_facebook_likes':'cast_total_facebook_likes'
    }
)

Проверяем ещё раз, чтобы всё было корректно

In [7]:
df.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes;'],
      dtype='object')

Теперь ещё раз выведем информацию о столбцах, чтобы увидеть возможные пропущенные значение:

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5025 non-null   object 
 1   director_name              4872 non-null   object 
 2   num_critic_for_reviews     4927 non-null   float64
 3   duration                   4959 non-null   float64
 4   director_facebook_likes    4872 non-null   float64
 5   actor_3_facebook_likes     4953 non-null   float64
 6   actor_2_name               4963 non-null   object 
 7   actor_1_facebook_likes     4968 non-null   float64
 8   gross                      4104 non-null   float64
 9   genres                     4974 non-null   object 
 10  actor_1_name               4968 non-null   object 
 11  movie_title                4974 non-null   object 
 12  num_voted_users            4974 non-null   float64
 13  cast_total_facebook_likes  4974 non-null   float

Как видим, пропущенные данные присутствуют в каждом поле, выведем более подробную информацию:

In [9]:
df.isna().sum()

color                         18
director_name                171
num_critic_for_reviews       116
duration                      84
director_facebook_likes      171
actor_3_facebook_likes        90
actor_2_name                  80
actor_1_facebook_likes        75
gross                        939
genres                        69
actor_1_name                  75
movie_title                   69
num_voted_users               69
cast_total_facebook_likes     69
actor_3_name                  90
facenumber_in_poster          82
plot_keywords                216
movie_imdb_link               69
num_user_for_reviews          87
language                      82
country                       72
content_rating               366
budget                       549
title_year                   174
actor_2_facebook_likes        80
imdb_score                    69
aspect_ratio                 389
movie_facebook_likes;         69
dtype: int64

Несмотря на то, что в гипотезах, которые будем проверять ниже, будет участвовать отнюнь не каждое поле, сделаем так, чтобы нигде не было пропущенных данных. Также убедимся, чтобы маркеры заполнялись соответственно типам данных полей:

In [10]:
values = {"color": "unknown", "director_name": "unknown", "num_critic_for_reviews": 0, "duration": 0,
          "director_facebook_likes":0, "actor_3_facebook_likes": 0, "actor_2_name": "unknown", "actor_1_facebook_likes": 0,
         "gross": 0, "genres": "unknown","actor_1_name": "unknown", "movie_title": "unknown", "num_voted_users":0,
         "cast_total_facebook_likes": 0, "actor_3_name": "unknown", "facenumber_in_poster": 0, "plot_keywords": "unknown", 
         "movie_imdb_link": "unknown", "num_user_for_reviews": 0, "language": "unknown", "country": "unknown", "content_rating": "unknown",
         "budget": 0, "title_year": 0, "actor_2_facebook_likes": 0, "imdb_score": 0, "aspect_ratio": 0, "movie_facebook_likes;": 0}
df = df.fillna(value=values)

Проверим, чтобы пропущенных данных не осталось:

In [11]:
df.isna().sum()

color                        0
director_name                0
num_critic_for_reviews       0
duration                     0
director_facebook_likes      0
actor_3_facebook_likes       0
actor_2_name                 0
actor_1_facebook_likes       0
gross                        0
genres                       0
actor_1_name                 0
movie_title                  0
num_voted_users              0
cast_total_facebook_likes    0
actor_3_name                 0
facenumber_in_poster         0
plot_keywords                0
movie_imdb_link              0
num_user_for_reviews         0
language                     0
country                      0
content_rating               0
budget                       0
title_year                   0
actor_2_facebook_likes       0
imdb_score                   0
aspect_ratio                 0
movie_facebook_likes;        0
dtype: int64

Проверим наличие дубликатов:

In [12]:
df.duplicated().sum()

45

Теперь отбросим дубликаты и обновим индексы:

In [13]:
df = df.drop_duplicates().reset_index(drop=True)

Проверим ещё раз:

In [14]:
df.duplicated().sum()

0

В гипотезах участвуют лишь поля gross, budget, imdb_score, director_name и country. Их и стоит проверять на неявные дубликаты, среди них дубликаты могут быть лишь среди полей типа данных "string" - director_name и contry. Посмотрим, на уникальные значения:

In [15]:
df['country'].unique()

array(['USA', 'UK', 'unknown', 'New Zealand', 'Canada', 'Australia',
       'Belgium', 'Japan', 'Germany', 'China', 'France', 'New Line',
       'Mexico', 'Spain', 'Hong Kong', 'Czech Republic', 'India',
       'Soviet Union', 'South Korea', 'Peru', 'Italy', 'Russia', 'Aruba',
       'Denmark', 'Libya', 'Ireland', 'South Africa', 'Iceland',
       'Switzerland', 'Romania', 'West Germany', 'Chile', 'Netherlands',
       'Hungary', 'Panama', 'Greece', 'Sweden', 'Norway', 'Official site',
       'Cambodia', 'Taiwan', 'Thailand', 'Slovakia', 'Bulgaria', 'Iran',
       'Poland', 'Georgia', 'Turkey', 'Nigeria', 'Brazil', 'Finland',
       'Bahamas', 'Argentina', 'Colombia', 'Israel', 'Egypt',
       'Kyrgyzstan', 'Indonesia', 'Pakistan', 'Slovenia', 'Afghanistan',
       'Dominican Republic', 'Cameroon', 'United Arab Emirates', 'Kenya',
       'Philippines'], dtype=object)

In [16]:
df['director_name'].unique()

array(['James Cameron', 'Gore Verbinski', 'Sam Mendes', ...,
       'Scott Smith', 'Benjamin Roberds', 'Daniel Hsia'], dtype=object)

Как видно выше, страны и имена режиссёров все пишутся с больших букв и без лишних пробелов, потому можно считать, что неявные дубликаты у полей, участвующих в исследовании, отсутствуют.

Теперь осталось преобразовать валовую и бюджетную стоимость в реальные суммы в долларах с точки зрения покупательной способности 2016 года.

In [24]:
from bs4 import BeautifulSoup
import requests

url = "https://www.usinflationcalculator.com/inflation/consumer-price-index-and-annual-percent-changes-from-1913-to-2008/"
r = requests.get(url)
data = r.text
soup = BeautifulSoup(data, "html.parser")

table = soup.find('table')
rows = table.tbody.findAll('tr')

years = []
cpis = []

for row in rows:
    year = row.findAll('td')[0].get_text()
    if year.isdigit() and int(year) < 2017:
        years.append(int(year))
        cpis.append(float(row.findAll('td')[13].get_text()))


cpi_table = pd.DataFrame({
    "year": years,
    "avg_annual_cpi": cpis
})

cpi_table.tail()

ConnectionError: HTTPSConnectionPool(host='www.usinflationcalculator.com', port=443): Max retries exceeded with url: /inflation/consumer-price-index-and-annual-percent-changes-from-1913-to-2008/ (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa57caf1ed0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))

In [63]:
def get_real_value(nominal_amt, old_cpi, new_cpi):
    real_value = (nominal_amt * new_cpi) / old_cpi
    return real_value


CPI = (cpi_table[cpi_table['year'] == 2016]['avg_annual_cpi'])
for x in CPI: CPI_2016 = x
# print(CPI_2016)

real_domestic_gross = []
real_budget_values = []

# должны преобразовать валовые и бюджетные значения 
# в реальное долларовое выражение 2016 г.

for index,row in df.iterrows():
    gross = row['gross']
    budget = row['budget']
    year = row['title_year']
    cp = (cpi_table[cpi_table['year'] == int(year)]['avg_annual_cpi'])

    for c in cp: cpi = c #равно единственному элементу Series cp

    real_gross = get_real_value(gross, cpi, CPI_2016)
    real_budget = get_real_value(budget, cpi, CPI_2016)
    real_domestic_gross.append(real_gross)
    real_budget_values.append(real_budget)
# print(real_domestic_gross)
df["real_domestic_gross"] = real_domestic_gross
df["real_budget"] = real_budget_values

profits = []
roi_vals = []

for index, row in df.iterrows():
    if row["real_domestic_gross"] !=0 and row["real_budget"] != 0:
        profit = row["real_domestic_gross"] - row["real_budget"]
        budget = row['real_budget']
        # конвертация ROI в проценты
        roi = (profit / budget) * 100
    else: profit = roi = 0

    profits.append(profit)
    roi_vals.append(roi)

df['profit'] = profits
df['roi'] = roi_vals

df.head(100)

color       director_name  \
0                                               Color       James Cameron   
1                                              Colour      Gore Verbinski   
2                                              Colour          Sam Mendes   
3                                               Color   Christopher Nolan   
4                                             unknown         Doug Walker   
5                                              Colour      Andrew Stanton   
6                                               Color           Sam Raimi   
7                                               Color        Nathan Greno   
8                                              Colour         Joss Whedon   
9                                              Colour         David Yates   
10                                              Color         Zack Snyder   
11                                              Color        Bryan Singer   
12                                              Color        Marc Forster   
13                                              Color      Gore Verbinski   
14                                              Color      Gore Verbinski   
15                                              color         Zack Snyder   
16                                              Color      Andrew Adamson   
17                                              Color         Joss Whedon   
18                                              Color        Rob Marshall   
19                                              Color    Barry Sonnenfeld   
20                                              Color       Peter Jackson   
21                                              color           Marc Webb   
22                                              Color        Ridley Scott   
23                                              Color       Peter Jackson   
24                                              Color         Chris Weitz   
25                                              Color       Peter Jackson   
26                                              Color       James Cameron   
27                                              Color       Anthony Russo   
28                                              Color          Peter Berg   
29                                              Color     Colin Trevorrow   
30                                              Color          Sam Mendes   
31                                              Color           Sam Raimi   
32                                              Color         Shane Black   
33                                              Color          Tim Burton   
34                                              Color        Brett Ratner   
35                                              Color         Dan Scanlon   
36                                              Color         Michael Bay   
37                                              Color         Michael Bay   
38                                              Color           Sam Raimi   
39                                              Color           Marc Webb   
40                                              Color     Joseph Kosinski   
41                                              Color       John Lasseter   
42                                              Color     Martin Campbell   
43                                              Color         Lee Unkrich   
44                                              Color                 McG   
45                                              Color           James Wan   
46                                              Color        Marc Forster   
47                                              Color        Bryan Singer   
48                                              Color         J.J. Abrams   
49                                              Color        Bryan Singer   
50                                              Color        Baz Luhrmann   
51                                          

# **Этап 3**. **Проверка гипотез**

# *Гипотеза 1: Окупшившеся фильмы имеют рейтинг IMDB больше 6*


Окупились - значит, значений поля profit - положительно

In [1]:
imdb_scores = [] # здесь будут храниться рейтинги всех окупившихся фильмов

for i in range(len(df)):  #remove the thing
    if df['profit'][i] > 0: imdb_scores.append(df['imdb_score'][i])

NameError: name 'df' is not defined

Давайте узнаем, сколько оценок, меньших или равных 6, в процентном соотношении.

In [65]:
wrong_counter = 0 #счётчик оценок, меньших или равных 6
correct_counter = 0 #счётчик оценок, больших 6
unknown_counter = 0 #счётчик неизвестных оценок (пропущенные значения)

for i in range(len(imdb_scores)):  #!!Сделать без цикла
    if imdb_scores[i] > 6: correct_counter += 1
    elif imdb_scores[i] == 0: unknown_counter += 1
    else: wrong_counter += 1
print(f'кол-во оценок, меньших или равных 6: {wrong_counter}')
print(f'кол-во оценок, больших 6: {correct_counter}')
print(f'кол-во оценок неизвестных оценок: {unknown_counter}')

кол-во оценок, меньших или равных 6: 498
кол-во оценок, больших 6: 1506
кол-во оценок неизвестных оценок: 0


Теперь вычислим процент всех неверных оценок, не согласованных с гипотезой:

In [66]:
print(wrong_counter / correct_counter *100) #edit

33.067729083665334


Таким образом, приблизительно 33% от всех окупаемых фильмов имеют рейтинг IMDB, меньший или равный 6. Получается, первая гипотеза правдива примерно на 100-33=67%.

# *Гипотеза 2: Более 50% всех фильмов, выпущенных США, окупились*

Более 50% - значит для того, чтобы гипотеза была верна, необходимо, что окупившихся фильмов США было больше половины от всех фильмов, выпущенных США

In [67]:
positive_profit_counter = 0
negative_profit_counter = 0
unknown_profit_counter = 0
for i in range(len(df)):
    if df['country'][i] == "USA": 
        if df['profit'][i] > 0:
            positive_profit_counter += 1
        elif df['profit'][i] == 0: unknown_profit_counter += 1
        else: negative_profit_counter += 1
            
print(f'кол-во окупившихся в США: {positive_profit_counter}')
print(f'кол-во неокупившихся в США: {negative_profit_counter}')
print(f'кол-во неизвестных полей profit в США: {unknown_profit_counter}')
print(f'общее кол-во фильмов США: {negative_profit_counter + positive_profit_counter + unknown_profit_counter}')

print(f'процент окупившихся от всех фильмов США без учёта неизвестных полей: {round(positive_profit_counter / (negative_profit_counter + positive_profit_counter) *100,2)}%')
print(f'процент окупившихся от всех фильмов США с учётом неизвестных полей: {round(positive_profit_counter / (negative_profit_counter + positive_profit_counter+unknown_profit_counter) *100,2)}%')
 

кол-во окупившихся в США: 1724
кол-во неокупившихся в США: 1291
кол-во неизвестных полей profit в США: 714
общее кол-во фильмов США: 3729
процент окупившихся от всех фильмов США без учёта неизвестных полей: 57.18%
процент окупившихся от всех фильмов США с учётом неизвестных полей: 46.23%


Как видим из последнего вывода:
- Без учета фильмов с неизвестным полем profit окупившихся фильмов больше 50%  
- C учётом фильмов с неизвестным полем profit окупившихся фильмов меньше 50% от всех фильмов- 46.23%. 
- Кол-во фильмов США с неизвестным полем profit достаточно, чтобы либо неокупившихся фильмов стало больше, либо наоборот - окупившихся. 

Можно, например, посчитать, сколько необходимо добавить из неизвестных в положительный профит, чтобы окупившихся фильмов стало больше половины:
Давайте высчитаем какого количества неизвестных полей профита не хватает положительным - окупаемым фильмам, чтобы процент был больше 50:

In [68]:
counter = 0 # счётчик кол-ва, которое необходимо добавить к положительным профитам от неизвестных
current_percentage = round(positive_profit_counter / (negative_profit_counter + positive_profit_counter+unknown_profit_counter) *100,2)
NEW_positive_profit_counter = positive_profit_counter # положительное кол-во профитов, которое будем инкрементировать

#убрать цикл
while current_percentage <= 50: # пишем условие, при котором мы в цикле инкрементируем кол-во полож профитов 
    NEW_positive_profit_counter += 1 # пока процент не превысит 50
    counter += 1
    current_percentage = round(NEW_positive_profit_counter / (negative_profit_counter + positive_profit_counter + unknown_profit_counter) *100,2)

print(f'кол-во которое необходимо добавить к положительным профитам от неизвестных, чтобы процент положительных профитов превысил 50: {counter}')
print(current_percentage)


кол-во которое необходимо добавить к положительным профитам от неизвестных, чтобы процент положительных профитов превысил 50: 141
50.01


Таким образом, можно сделать вывод, что из 714 фильмов США с неизвестным полем "profit" достаточно только чтобы 141 фильм с неизвестным полем "profit" перешёл в положительный профит - окупился, чтобы окупившихся фильмов США было больше половины.  

Итак, подводя итог:
- без учёта фильмов с неизвестными полями - гипотеза верна
- с учётом фильмов с неизвестными полями нельзя точно сказать - гипотеза и не верна и не неверна.

Таким образом, гипотеза верна частично.

# *Гипотеза 3: Все фильмов от Христофера Нолана имеют IMDB рейтинг больше 7*

In [69]:
for i in range(len(df)):
    if df['director_name'][i] == "Christopher Nolan": print(df['imdb_score'][i])

8.5
9.0
8.6
8.8
8.3
7.2
8.5
8.5


Как видим по выводу, в самом деле: рейтинги всех фильмов Христофера Нолана на сервисе IMDB больше 7. Таким образом гипотеза оказалась, на удачу, приятно проста.

# Заключение

Были проверены следующие три гипотезы:
1) Все фильмы с рейтингом IMDB больше 6 окупились
2) более 50% всех фильмов, выпущенных США, окупились
3) все фильмов от Христофера Нолана имеют IMDB рейтинг больше 7

*В первой гипотезе*: приблизительно 33% от всех окупаемых фильмов имеют рейтинг IMDB, меньший или равный 6. Таким образом получилось, что первая гипотеза правдива на 100-33=67%.

*Во второй гипотезе*: без учёта фильмов с неизвестными полями - гипотеза верна,
с учётом фильмов с неизвестными полями нельзя точно сказать - гипотеза и не верна и не неверна.
Таким образом мы пришли к выводу, что гипотеза верна частично.

*В третьей гипотезе*: всё однозначно: все фильмы Христофера Нолана на сервисе IMDB имеют рейтинг больше 7 - гипотеза верна.

In [70]:
df.to_csv('../Lab1_Gunter_FIT-221.csv')

# Доп. задание с проверкой первой гипотезы из примера

*Гипотеза 1: Активность пользователей различается в зависимости от дня недели и от города к городу*

Возьмём датасет из примера и добавим немного своего:

In [5]:
data = {"userid": ['FFB692EC', '55204538', '20EC38', 'A3DD03C9', 'E2DC1FAE', '842029A1', '4CB90AA5', 'F03E1C1F', '8FA1D3BE', 'E772D5C0', 'D7D2D9CC',"SDF12349", "1234BEDI"],
        "track": ["Kamigata To Boots", 'Delayed Because of Accident', 'Funiculì funiculà', 'Dragons in the Sunset', 'Soul People', 'Chains', 'True',
                 'Feeling This Way', "L’estate", 'Pessimist', "That's the way", "Through the darkness", "How to live"],
        "artist": ["The Mass Missile", "Andreas Rönnberg", "Mario Lanza", "Fire + Ice", "Space Echo", "Obladaet", "Roman Messer", "Polina Griffith",
                  "Julia Dalia", "unknown", "Ivan Durov", "Imagine Dragons", "NF"],
        "genre": ["rock", "rock", "pop","folk", "dance","rusrap","dance","dance","ruspop","dance", "hiphop","rock","hiphop"],
        "city": ["Shelbyville", "Springfield", "Shelbyville", "Shelbyville", "Springfield", "Shelbyville", "Springfield","Springfield","Springfield", "Shelbyville", 
                 "Shelbyville", "Springfield", "Shelbyville"],
        "time": ["20:28:33", "14:07:09", "20:58:07", "08:37:09", "08:34:34", "13:09:41", "13:00:07", "20:47:49", "09:17:40", "21:20:49", "20:19:08", "02:06:10","12:48:19"],
         "day": ["Wednesday", "Friday", "Wednesday", "Monday", "Monday", "Friday", "Wednesday", "Wednesday", "Friday", "Wednesday", "Wednesday", "Wednesday", "Wednesday"]} 

df1 = pd.DataFrame(data)
df1

userid                        track            artist   genre  \
0   FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1   55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2     20EC38            Funiculì funiculà       Mario Lanza     pop   
3   A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4   E2DC1FAE                  Soul People        Space Echo   dance   
5   842029A1                       Chains          Obladaet  rusrap   
6   4CB90AA5                         True      Roman Messer   dance   
7   F03E1C1F             Feeling This Way   Polina Griffith   dance   
8   8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9   E772D5C0                    Pessimist           unknown   dance   
10  D7D2D9CC               That's the way        Ivan Durov  hiphop   
11  SDF12349         Through the darkness   Imagine Dragons    rock   
12  1234BEDI                  How to live                NF  hiphop   

           city      time        day  
0   Shelbyville  20:28:33  Wednesday  
1   Springfield  14:07:09     Friday  
2   Shelbyville  20:58:07  Wednesday  
3   Shelbyville  08:37:09     Monday  
4   Springfield  08:34:34     Monday  
5   Shelbyville  13:09:41     Friday  
6   Springfield  13:00:07  Wednesday  
7   Springfield  20:47:49  Wednesday  
8   Springfield  09:17:40     Friday  
9   Shelbyville  21:20:49  Wednesday  
10  Shelbyville  20:19:08  Wednesday  
11  Springfield  02:06:10  Wednesday  
12  Shelbyville  12:48:19  Wednesday

In [10]:
import numpy as np
#исправить таблицу чтобы дни недели были в строку (столбцы) - DONE
pivot = pd.pivot_table(df1,
               index= ["city", ], 
               columns=["day"],
               values = ["track", ],
               aggfunc = ["count"])
pivot

count                 
             track                 
day         Friday Monday Wednesday
city                               
Shelbyville      1      1         5
Springfield      2      1         3

Анализируя полученные данные, видим, что, на самом деле, активность слушателей варьируется в зависимости от дня недели: так в городе Спрингфилд в понедельник слушает 1 пользователь, а в пятницу и среду 2 и 3 соответственно.
Аналогично тип города также влияет на активность: в Шелбивилле в понедельник, среду и пятницу 1 1 и 5 слушателя соответственно, а в Спригфилде - в понедельник, среду и пятницу 1 соответственно, как уже было упомянуто выше.

Таким образом, гипотеза верна.